solver/src/main.rs

```diff
@@ -300,48 +320,35 @@ async fn main() {
 }
 
 async fn build_amm_artifacts(
-    sources: Vec<BaselineSources>,
-    chain_id: u64,
+    sources: &HashMap<BaselineSources, Arc<PoolCache>>,
     settlement_contract: contracts::GPv2Settlement,
     base_tokens: HashSet<H160>,
     web3: shared::Web3,
 ) -> Vec<UniswapLikeLiquidity> {
     let mut res = vec![];
-    for source in sources {
-        match source {
-            BaselineSources::Uniswap => {
-                let router = contracts::UniswapV2Router02::deployed(&web3)
-                    .await
-                    .expect("couldn't load deployed uniswap router");
-                let factory = contracts::UniswapV2Factory::deployed(&web3)
-                    .await
-                    .expect("couldn't load deployed uniswap router");
-                let pair_provider = Arc::new(UniswapPairProvider { factory, chain_id });
-                res.push(UniswapLikeLiquidity::new(
-                    IUniswapLikeRouter::at(&web3, router.address()),
-                    pair_provider.clone(),
-                    settlement_contract.clone(),
-                    base_tokens.clone(),
-                    web3.clone(),
-                ));
-            }
-            BaselineSources::Sushiswap => {
-                let router = contracts::SushiswapV2Router02::deployed(&web3)
-                    .await
-                    .expect("couldn't load deployed sushiswap router");
-                let factory = contracts::SushiswapV2Factory::deployed(&web3)
-                    .await
-                    .expect("couldn't load deployed sushiswap router");
-                let pair_provider = Arc::new(SushiswapPairProvider { factory });
-                res.push(UniswapLikeLiquidity::new(
-                    IUniswapLikeRouter::at(&web3, router.address()),
-                    pair_provider.clone(),
-                    settlement_contract.clone(),
-                    base_tokens.clone(),
-                    web3.clone(),
-                ));
-            }
-        }
+    if let Some(pool_fetcher) = sources.get(&BaselineSources::Uniswap) {
+        let router = contracts::UniswapV2Router02::deployed(&web3)
+            .await
+            .expect("couldn't load deployed uniswap router");
+        res.push(UniswapLikeLiquidity::new(
+            IUniswapLikeRouter::at(&web3, router.address()),
+            settlement_contract.clone(),
+            base_tokens.clone(),
+            web3.clone(),
+            pool_fetcher.clone(),
+        ));
+    }
+    if let Some(pool_fetcher) = sources.get(&BaselineSources::Sushiswap) {
+        let router = contracts::SushiswapV2Router02::deployed(&web3)
+            .await
+            .expect("couldn't load deployed sushiswap router");
+        res.push(UniswapLikeLiquidity::new(
+            IUniswapLikeRouter::at(&web3, router.address()),
+            settlement_contract.clone(),
+            base_tokens.clone(),
+            web3.clone(),
+            pool_fetcher.clone(),
+        ));
```

Comment:
Given that we are no longer matching on the source here, I would think that when introducing a new source, one could easily forget to include the necessary code here and be confused as to why their code isn't working. Is there any way we could somehow circumvent this or is my concern not actually an issue?

Reply:
Changed code to match